compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -305,7 +317,7 @@ impl<'a> Parser<'a> {
         // Try to recover from use of `+` with incorrect priority.
         self.maybe_report_ambiguous_plus(allow_plus, impl_dyn_multi, &ty);
         self.maybe_recover_from_bad_type_plus(allow_plus, &ty)?;
-        let ty = self.maybe_recover_from_question_mark(ty, is_as_cast);
+        let ty = self.maybe_recover_from_question_mark(ty, recover_question_mark);
```

Comment:
If further problems are encountered in the future, then the appropriate change will be to change this to return the specific `Err()` instead of allowing successful subsequent parses. That'd be a shame, but it'd be a straightforward change (and don't envision it being a problem anywhere else other than exprs and macros, which are now both handled).